### ISSUE TITLE
Integrations tests are slow

### ISSUE THREAD

**Issue by naegelejd** [gadgetron/gadgetron#101]
List of causes, ordered by difficulty to solve (easiest to hardest):

1. Unnecessary `time.sleep(2)` call [here](https://github.com/gadgetron/gadgetron/blob/cc7efc2689bb85ec850a66e56d0c3be944441297/test/integration/run_gadgetron_test.py#L80)
1. Unnecessary `time.sleep(1)` call [here](https://github.com/gadgetron/gadgetron/blob/cc7efc2689bb85ec850a66e56d0c3be944441297/test/integration/run_gadgetron_test.py#L269)
1. `gadgetron_info` is called using subprocess for each task. 5 regular expressions are compiled per test to parse the output of `gadgetron_info`
1. A new gadgetron instance is spawned for each integration test. This means that long-running Gadgetron instances are not tested for memory leaks, etc.

1 and 2 are easy to fix. I propose fixing 3 by splitting out the system-checking code so that `gadgetron_info` is executed once if `run_all_tests.py` is run or if `run_gadgetron_test.py` is run in a standalone fashion. 4 is debatable but I believe I raise a good issue about not testing "long-running" instances.

**Comment by hansenms** [comment]
@naegelejd, does 1,2, and 3 really add any significant amount of time to the overall time that it takes to run all the tests?

I agree that our current tests do not test if for leaks in long running Gadgetrons, however, if we do not spawn a new gadgetron instance for each test, it is also hard to see what specific case is failing if there is a leak. We should really take some cases and run them many times against the same instance, but that would take even more time. I agree there is some need for this type of testing, but I don't think that running all tests against a single instance is the solution.

**Comment by inati** [comment]
30 tests x 3s = 1.5 min doing nothing

**Comment by naegelejd** [comment]
If the integration tests take 10 minutes, 15% of that time is calls to `time.sleep()` that serve virtually no purpose. I only included these for completeness though. I understand the need to spawn individual Gadgetron instances so I don't have a "solution" to my complaint. My intention was to get approval for cleaning up the integration tests in general, but I admit I haven't made a great case yet.

**Comment by hansenms** [comment]
Please feel free to do any cleaning up that you would like. And any speed up would be greatly appreciated, but....

Be aware that the sleep in 2 is only used for chroot testing, and is absolutely needed to make sure that things are settled before starting next test. I believe the first sleep is needed too. I can't remember the issue, but probably the following mkdirs would not work if called right after the folder is deleted. However, that could be fixed by only creating it when needed.

In any case, the assertion that it is spent doing "nothing" is incorrect. The sleeps have been added because they were needed (at some point in time). Can they be removed or shortened, maybe. But still they are not doing. 

So I think that the real expense here (for the standard case) is 30 x 2s or 1 minute our of 10. You may be able to cut that in half, but we would not be talking about any dramatic changes here. So lets say you can cut it down to 9.5 minutes or reduce by 5%, it would still be slow, right? That being said, 5% is 5% so by all means, knock yourself out;)

**Comment by naegelejd** [comment]
I'd love to do some cleanup and potentially speed some things up.

To be clear, both `sleep` calls I pointed out do virtually nothing, and they did nothing when they were added (see 7bc0418f and 898fc3aa). `shutil.rmtree` is a blocking call, and nothing else executes in the interpreter while that's happening. The next calls is `os.makedirs`, which re-creates the deleted directory, but that code does not, and will never, run before the `shutil.rmtree` finishes. The `sleep` after calling `subprocess.call` to kill the running gadgetron does nothing. `subprocess.call` does not return until the subprocess has finished completely and returned a return code. From the [docs](https://docs.python.org/2/library/subprocess.html#subprocess.call): "Run the command described by args. Wait for command to complete, then return the returncode attribute."

No big deal, just clarifying.
